archive.go

```diff
@@ -2,157 +2,88 @@ package k6deps
-
-		switch header.Typeflag {
-		case tar.TypeDir:
-			if err := os.MkdirAll(target, 0o750); err != nil {
-				return err
-			}
+		if header.Typeflag != tar.TypeReg || !shouldProcess(header.Name) {
+			continue
+		}
 
-		case tar.TypeReg:
-			if shouldSkip(target) {
-				continue
-			}
+		content := &bytes.Buffer{}
+		if _, err := io.CopyN(content, reader, maxFileSize); err != nil && !errors.Is(err, io.EOF) {
+			return nil, err
+		}
 
-			file, err := os.OpenFile(filepath.Clean(target), os.O_CREATE|os.O_RDWR, os.FileMode(header.Mode)) //nolint:gosec
+		// if the file is metadata.json, we extract the dependencies from the env
+		if header.Name == "metadata.json" {
+			analyzer, err := analizeMetadata(content.Bytes())
 			if err != nil {
-				return err
-			}
-
-			if _, err := io.CopyN(file, reader, maxFileSize); err != nil && !errors.Is(err, io.EOF) {
-				return err
-			}
-
-			if err = file.Close(); err != nil {
-				return err
-			}
-
-		// if it is a link or symlink, we copy the content of the linked file to the target
-		// we assume the linked file was already processed and exists in the directory.
-		case tar.TypeLink, tar.TypeSymlink:
-			if shouldSkip(target) {
-				continue
+				return nil, err
 			}
+			analyzers = append(analyzers, analyzer)
+			continue
+		}
 
-			linkedFile := filepath.Join(dir, filepath.Clean(filepath.FromSlash(header.Linkname)))
-			if err := followLink(linkedFile, target); err != nil {
-				return err
-			}
+		// analize the file content as an script
+		target := filepath.Clean(filepath.FromSlash(header.Name))
+		src := Source{
+			Name:     target,
+			Contents: content.Bytes(),
 		}
+
+		analyzers = append(analyzers, scriptAnalyzer(src))
 	}
 }
 
-// indicates if the file should be skipped during extraction
-// we skip csv files and .json except metadata.json
-func shouldSkip(target string) bool {
+// indicates if the file should be processed during extraction
+func shouldProcess(target string) bool {
 	ext := filepath.Ext(target)
-	return ext == ".csv" || (ext == ".json" && filepath.Base(target) != "metadata.json")
+	return slices.Contains([]string{".js", ".ts"}, ext) || slices.Contains([]string{"metadata.json", "data"}, target)
 }
 
-//nolint:forbidigo
-func followLink(linkedFile string, target string) error {
-	source, err := os.Open(filepath.Clean(linkedFile))
-	if err != nil {
-		return err
-	}
-	defer source.Close() //nolint:errcheck
-
-	// we need to get the lined file info to create the target file with the same permissions
-	info, err := source.Stat()
-	if err != nil {
-		return err
-	}
-
-	file, err := os.OpenFile(target, os.O_CREATE|os.O_WRONLY, info.Mode()) //nolint:gosec
-	if err != nil {
-		return err
+// analizeMetadata extracts the dependencies from the metadata.json file
+func analizeMetadata(content []byte) (analyzer, error) {
+	metadata := archiveMetadata{}
+	if err := json.Unmarshal(content, &metadata); err != nil {
```

Comment:
If I'm not mistaken, we could use `json.NewDecoder` and then there is no need to pass bytes, but we could use `io.Reader` directly here